boards/imix/src/main.rs

```diff
@@ -246,360 +247,64 @@ pub unsafe fn reset_handler() {
-        [
-            &sam4l::adc::CHANNEL_AD1, // AD0
-            &sam4l::adc::CHANNEL_AD2, // AD1
-            &sam4l::adc::CHANNEL_AD3, // AD2
-            &sam4l::adc::CHANNEL_AD4, // AD3
-            &sam4l::adc::CHANNEL_AD5, // AD4
-            &sam4l::adc::CHANNEL_AD6, // AD5
-        ]
-    );
-    let adc = static_init!(
-        capsules::adc::Adc<'static, sam4l::adc::Adc>,
-        capsules::adc::Adc::new(
-            &mut sam4l::adc::ADC0,
-            adc_channels,
-            &mut capsules::adc::ADC_BUFFER1,
-            &mut capsules::adc::ADC_BUFFER2,
-            &mut capsules::adc::ADC_BUFFER3
-        )
-    );
-    sam4l::adc::ADC0.set_client(adc);
-
-    // # GPIO
-    // set GPIO driver controlling remaining GPIO pins
-    let gpio_pins = static_init!(
-        [&'static sam4l::gpio::GPIOPin; 7],
-        [
-            &sam4l::gpio::PC[31], // P2
-            &sam4l::gpio::PC[30], // P3
-            &sam4l::gpio::PC[29], // P4
-            &sam4l::gpio::PC[28], // P5
-            &sam4l::gpio::PC[27], // P6
-            &sam4l::gpio::PC[26], // P7
-            &sam4l::gpio::PA[20], // P8
-        ]
-    );
-
-    let gpio = static_init!(
-        capsules::gpio::GPIO<'static, sam4l::gpio::GPIOPin>,
-        capsules::gpio::GPIO::new(gpio_pins)
-    );
-    for pin in gpio_pins.iter() {
-        pin.set_client(gpio);
-    }
-
-    // # LEDs
-    let led_pins = static_init!(
-        [(&'static sam4l::gpio::GPIOPin, capsules::led::ActivationMode); 2],
-        [
-            (
-                &sam4l::gpio::PC[22],
-                capsules::led::ActivationMode::ActiveHigh
-            ),
-            (
-                &sam4l::gpio::PC[10],
-                capsules::led::ActivationMode::ActiveHigh
-            ),
-        ]
-    );
-    let led = static_init!(
-        capsules::led::LED<'static, sam4l::gpio::GPIOPin>,
-        capsules::led::LED::new(led_pins)
-    );
-
-    // # BUTTONs
-
-    let button_pins = static_init!(
-        [(&'static sam4l::gpio::GPIOPin, capsules::button::GpioMode); 1],
-        [(
-            &sam4l::gpio::PC[24],
-            capsules::button::GpioMode::LowWhenPressed
-        )]
-    );
-
-    let button = static_init!(
-        capsules::button::Button<'static, sam4l::gpio::GPIOPin>,
-        capsules::button::Button::new(button_pins, kernel::Grant::create())
-    );
-    for &(btn, _) in button_pins.iter() {
-        btn.set_client(button);
-    }
-
-    let crc = static_init!(
-        capsules::crc::Crc<'static, sam4l::crccu::Crccu<'static>>,
-        capsules::crc::Crc::new(&mut sam4l::crccu::CRCCU, kernel::Grant::create())
-    );
+    let adc = AdcComponent::new().finalize();
+    let gpio = GpioComponent::new().finalize();
+    let led = LedComponent::new().finalize();
+    let button = ButtonComponent::new().finalize();
+    let crc = CrcComponent::new().finalize();
 
-    rf233_spi.set_client(rf233);
+    // Can this initialize be pushed earlier, or into component? -pal
```

Comment:
On this note, could the initialization/finalization of RF233 and rf233 spi be pushed into the initialization/finalization for the radioComponent? Given that there wouldn't be a scenario where the Imix radioComponent is used without these, it seems like just having what is currently line 304 handle all of that might be the desired interface (though of course the RadioComponet::new() would have to take in the parameters required for RF233Component::new() and SpiComponent::new()

Perhaps this doesn't make sense, if so, let me know.

Reply:
I think that's another way to do it. I kept them separate for the case when someone wants to use the radio but not 6lowpan. I don't have a great answer for the case when multiple Components all initialize the same subsystem (e.g., suppose 6lowpan initializes Radio, but you include both the Radio and the 6lowpan Components: I'm sure hard-to-debug errors will occur). So I chose to keep the dependencies explicit.